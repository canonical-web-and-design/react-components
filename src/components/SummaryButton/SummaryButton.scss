@import "~vanilla-framework/scss/settings";

.p-summary__text {
  display: inline-block;
  margin-right: $sph-inner--small;
}
