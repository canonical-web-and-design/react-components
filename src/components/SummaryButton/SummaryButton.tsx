import React, { SyntheticEvent } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

import ActionButton from "../ActionButton";

type Props = {
  className?: string;
  label: string;
  onClick: (event: SyntheticEvent) => void;
  summary?: string;
  isLoading?: boolean;
};

const SummaryButton = ({
  className,
  isLoading,
  summary,
  label,
  onClick,
}: Props): JSX.Element => (
  <small className={classNames("u-text--muted", className)}>
    {summary && (
      <>
        <span>{summary}</span>{" "}
      </>
    )}
    {onClick && (
      <ActionButton
        appearance="neutral"
        className="is-small is-dense"
        onClick={onClick}
        loading={isLoading}
        disabled={isLoading}
      >
        {label}
      </ActionButton>
    )}
  </small>
);

SummaryButton.propTypes = {
  className: PropTypes.string,
  summary: PropTypes.string,
  label: PropTypes.string,
  onClick: PropTypes.func,
  isLoading: PropTypes.bool,
};

export default SummaryButton;
