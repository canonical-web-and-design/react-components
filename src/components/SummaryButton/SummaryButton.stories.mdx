import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs/blocks";
import { useState } from "react";

import SummaryButton from "./SummaryButton";

<Meta title="SummaryButton" component={SummaryButton} />

## SummaryButton

`SummaryButton` is a component to render a short summary with an action button. It's initial use case is a summary of table contents with action to load more items.

### Props

<ArgsTable of={SummaryButton} />

### Default

export const doNothing = () => {};

<Canvas>
  <Story name="Default">
    <SummaryButton summary="Showing 15 out of 100 items." label="Show more" onClick={doNothing} />
  </Story>
</Canvas>

### Action only

Summary text is optional.

<Canvas>
  <Story name="Action only">
    <SummaryButton label="Show more items" onClick={doNothing} />
  </Story>
</Canvas>

### Summary only

Action is optional as well. When `onClick` prop is not provided component will render just the summary text.

<Canvas>
  <Story name="Summary only">
    <SummaryButton summary="Showing 100 out of 100 items." label="Show more items" />
  </Story>
</Canvas>

### Loading

When data is loaded asynchronusly `isLoading` prop should be set to `true` to shows a spinner animation in place of the action button.

<Canvas>
  <Story name="Loading">
    <SummaryButton summary="Showing 15 out of 100 items." label="Show more" isLoading onClick={doNothing} />
  </Story>
</Canvas>
