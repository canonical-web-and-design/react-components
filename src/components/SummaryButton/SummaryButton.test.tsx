import { shallow } from "enzyme";
import React from "react";

import SummaryButton from "./SummaryButton";

describe("<SummaryButton />", () => {
  // snapshot tests
  it("renders and matches the snapshot", () => {
    const component = shallow(
      <SummaryButton
        summary="Showing some items"
        label="Show more"
        onClick={() => false}
      />
    );

    expect(component).toMatchSnapshot();
  });

  // unit tests
  it("renders a spinner icon", () => {
    const component = shallow(
      <SummaryButton label="Show more" onClick={() => false} isLoading />
    );

    expect(component.find("ActionButton").first().prop("loading")).toBeTruthy();
  });

  it("can handle click events", () => {
    const onClick = jest.fn();
    const component = shallow(
      <SummaryButton label="Show more" onClick={onClick} />
    );

    component.find("ActionButton").first().simulate("click");
    expect(onClick).toHaveBeenCalled();
  });
});
