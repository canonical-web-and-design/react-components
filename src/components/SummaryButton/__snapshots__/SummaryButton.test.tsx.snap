// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SummaryButton /> renders and matches the snapshot 1`] = `
<small
  className="u-text--muted"
>
  <span>
    Showing some items
  </span>
   
  <ActionButton
    appearance="neutral"
    className="is-small is-dense"
    onClick={[Function]}
  >
    Show more
  </ActionButton>
</small>
`;
